pinot-common/src/main/java/com/linkedin/pinot/common/utils/FileUploadUtils.java

```diff
@@ -77,43 +82,51 @@ public EntityEnclosingMethod forUri(String uri) {
   public static int sendFile(final String host, final String port, final String path, final String fileName,
       final InputStream inputStream, final long lengthInBytes, SendFileMethod httpMethod) {
     EntityEnclosingMethod method = null;
-    try {
-      method = httpMethod.forUri("http://" + host + ":" + port + "/" + path);
-      Part[] parts = {
-          new FilePart(fileName, new PartSource() {
-            @Override
-            public long getLength() {
-              return lengthInBytes;
-            }
-
-            @Override
-            public String getFileName() {
-              return "fileName";
-            }
-
-            @Override
-            public InputStream createInputStream() throws IOException {
-              return new BufferedInputStream(inputStream);
-            }
-          })
-      };
-      method.setRequestEntity(new MultipartRequestEntity(parts, new HttpMethodParams()));
-      FILE_UPLOAD_HTTP_CLIENT.executeMethod(method);
-      if (method.getStatusCode() >= 400) {
-        String errorString = "POST Status Code: " + method.getStatusCode() + "\n";
-        if (method.getResponseHeader("Error") != null) {
-          errorString += "ServletException: " + method.getResponseHeader("Error").getValue();
+    final BufferedInputStream bufferedInputStream = new BufferedInputStream(inputStream);
+    for (int numRetries = 0; ; numRetries++) {
+      try {
+        if (numRetries > 0) {
+          bufferedInputStream.reset();
+        }
+        bufferedInputStream.mark(0);
+        method = httpMethod.forUri("http://" + host + ":" + port + "/" + path);
+        Part[] parts = {new FilePart(fileName, new PartSource() {
+          @Override
+          public long getLength() {
+            return lengthInBytes;
+          }
+
+          @Override
+          public String getFileName() {
+            return fileName;
+          }
+
+          @Override
+          public InputStream createInputStream()
+              throws IOException {
+            return new CloseShieldInputStream(bufferedInputStream);
+          }
+        })};
+        method.setRequestEntity(new MultipartRequestEntity(parts, new HttpMethodParams()));
+        FILE_UPLOAD_HTTP_CLIENT.executeMethod(method);
+        if (method.getStatusCode() >= 400) {
+          String errorString = "POST Status Code: " + method.getStatusCode() + "\n";
+          if (method.getResponseHeader("Error") != null) {
+            errorString += "ServletException: " + method.getResponseHeader("Error").getValue();
+          }
+          throw new HttpException(errorString);
+        }
+        return method.getStatusCode();
+      } catch (Exception e) {
+        LOGGER.error("Caught exception while sending file: {}", fileName, e);
```

Comment:
Better to log a warning here, and log error when the numRetries already reach MAX_RETRIES.